spacy/tokens/doc.pyx

```diff
@@ -1462,6 +1462,146 @@ cdef class Doc:
+            self.push_back(lex, has_space)
+
+        # Set remaining token-level attributes via Doc.from_array().
+        if token_annotations[HEAD] is not None:
+            token_annotations[HEAD] = [
+                head - i if head is not None else 0 for i, head in enumerate(token_annotations[HEAD])
+            ]
+        if token_annotations[DEP] is not None:
+            deps = []
+            for dep in token_annotations[DEP]:
+                if dep is None:
+                    raise ValueError(Errors)
+                deps.append(dep)
+            token_annotations[DEP] = deps if len(deps) else None
+
+        if token_annotations[DEP] and not token_annotations[HEAD]:
+            token_annotations[HEAD] = [0] * len(token_annotations[DEP])
+        if token_annotations[HEAD] and not token_annotations[HEAD]:
+            raise ValueError(Errors.E1017)
+        if token_annotations[POS] is not None:
+            for pp in set(token_annotations[POS]):
+                if pp not in parts_of_speech.IDS:
+                    raise ValueError(Errors.E1021.format(pp=pp))
+
+        attrs = []
+        annotations = [token_annotations[key] for key in annot_types]
+        possible_headings = [POS, HEAD, DEP, LEMMA, TAG, MORPH]
+        for a, annot in enumerate(annotations):
+            if annot is not None:
+                if len(annot) != len(words):
+                    raise ValueError(Errors.E189)
+                attrs.append(possible_headings[a])
+
+        # If there are any other annotations, set them.
+        if attrs:
+            array = self.to_array(attrs)
+            if array.ndim == 1:
+                array = numpy.reshape(array, (array.size, 1))
+            j = 0
+
+            for annot in annotations:
+                if annot:
+                    if annot is token_annotations[HEAD]:
+                        for i in range(len(words)):
+                            array[i, j] = annot[i]
+                    elif annot is token_annotations[MORPH]:
+                        for i in range(len(words)):
+                            array[i, j] = self.vocab.morphology.add(token_annotations[MORPH][i])
+                    else:
+                        for i in range(len(words)):
+                            array[i, j] = self.vocab.strings.add(annot[i])
+                    j += 1
+            self.from_array(attrs, array)
+
+        ### Span/document properties ###
+
+        # Complement other document-level properties (cats, spans, ents).
+        self.cats = doc_json.get("cats", self.cats)
+
+        # Set sentence boundaries, if dependency parser not available but sentences are specified in JSON.
+        if not self.has_annotation("DEP"):
+            for sent in doc_json.get("sents", {}):
+                for t in self.char_span(sent["start"], sent["end"]):
+                    self[t.i].is_sent_start = t.i == 0
+
+        if "spans" in doc_json:
+            for span_group in doc_json.get("spans", {}):
```

Comment:
```suggestion
            for span_group in doc_json.get["spans"]:
```

Or the spans/ents could be unindented with `for span_group in doc_json.get("spans", {})`?

Reply:
Did you mean `doc_json.get("spans")` or `doc_json["spans"]`?

Reply:
I meant either:

```python
if "spans" in doc_json:
    for span_group in doc_json["spans"]:
```

or

```python
for span_group in doc_json.get("spans", {}):
```

(and unindent one level)

The second version looks more like `"cats"` and `_`, so I'd suggest that?